natiami/assignment1/Q1.cpp

```diff
@@ -2,31 +2,56 @@
 #include<string>
 #include<vector>
 #include<fstream>
-#include <algorithm>
-#include <array> 
+#include<algorithm>
+#include<array> 
 
 using namespace std;
 
+int letterToIndex(char letter){
+	// substracting 65 to start indexing from 0, integer value of A symbol is 65
+	return int(letter) - 65; 
+}
+
+int indexToLowerCase(int index){
+	return index + ('a' - 'A');
+}
+
+bool checkAllZeros(vector<int> letters_histogram){
+	return all_of(letters_histogram.begin(), letters_histogram.end(), [](int letter_count){return letter_count == 0; });
+}
+
 bool checkAnagrams(string str1, string str2, bool caseSensitive){
-	vector<int> letters(52, 0); // array for letters
+	int nr_symbols = 58; // number of symbols from A to z
+	vector<int> letters_histogram(nr_symbols, 0);
 
-	if (str1.length() != str2.length()) return false; // if size differs return false right away
+	if (str1.length() != str2.length()) return false; // Different size words cannot be anagrams
 
 	for (size_t i = 0; i < str1.length(); i++)
 	{
-		letters[int(str1[i]) - 65]++;
-		letters[int(str2[i]) - 65]--;
+		letters_histogram[letterToIndex(str1[i])]++;
+		letters_histogram[letterToIndex(str2[i])]--;
 	}
 
-	// If algorithm is not case sensitive
+	// Case insensitive anagrams check
 	if (!caseSensitive){
-		for (size_t i = 0; i < letters.size()/2; i++)
-			if (letters[i] != 0 && letters[i + 32] == 0) return false;
-		return true;
+		const int A_index = 0;
+		const int Z_index = 26; // english language has 26 letters
+
+		for (size_t i = A_index; i < Z_index; i++){
+			if (letters_histogram[i] != 0){
+				if (abs(letters_histogram[i]) != abs(letters_histogram[indexToLowerCase(i)]))
+					return false;
+				else{
+					letters_histogram[indexToLowerCase(i)] = 0;
+					letters_histogram[i] = 0;
+				}
+			}
+		}
+		return checkAllZeros(letters_histogram);

```

Comment:
Actually, I added this part because I only compare histogram of upper letter with lower letters but if words include all lower case letters without this part algorithm will classifies aaaa and bbbb as anagrams. Because all upper case letters are 0 and it will not cause conflict

Reply:
If the input is "aaaa", "bbbb",
then the histogram is:
a:4, b:-4, A:0, B:0
so `abs(histogram['A']) == abs(histogram['a'])` (which is `0 == 4`), so it will return false and not classify as anagrams. 
Is that right or the histogram will be different?